### ISSUE TITLE
qiskit/visualization/counts_visualization.py needs testing

### ISSUE THREAD

**Issue by 1ucian0** [Qiskit/qiskit-terra#2307]
The code in `qiskit/visualization/counts_visualization.py` needs testing. Current coverage is:

file | statements | missing | coverage
-- | -- | -- | --
qiskit/visualization/counts_visualization.py | 98 | 87 | 11%

(This is a follow up from https://github.com/Qiskit/qiskit-terra/issues/300#issuecomment-489324685)

**Comment by 1ucian0** [comment]
This is still true: https://coveralls.io/builds/24960860/source?filename=qiskit/visualization/counts_visualization.py

**Comment by 1ucian0** [comment]
`on hold` until #2968 is decided.

**Comment by tgag17** [comment]
Hey! Is the issue still open?

**Comment by 1ucian0** [comment]
it is!

Removing `on hold`. The testing should be on binder, like in https://github.com/Qiskit/qiskit-terra/pull/4544x

**Comment by pawmasz** [comment]
Is this issue still open? Can I take a look at it?

**Comment by javabster** [comment]
As there are currently no other snapshot tests for counts (only circuits) I suggest creating a new file for count visualisation tests.

More specifically, I suggest the following steps for working on this issue:
1. Create new file `qiskit-terra/test/ipynb/mpl/test_counts_matplotlib_drawer.py`
2. Add new tests to this file (use `test_counts_matplotlib_drawer.py` as a guide)
3. Update `qiskit-terra/test/ipynb/mpl_tester.ipynb` to also run tests from the `test_counts_matplotlib_drawer.py`
4. Follow the instructions in the [here](https://github.com/Qiskit/qiskit-terra/blob/main/CONTRIBUTING.md#snapshot-testing-for-visualizations) about how to run the snapshot tests and update the reference images

**Comment by javabster** [comment]
@tgag17 are you still interested in working on this?

**Comment by tgag17** [comment]
Hi! Yes, I would like to work on it. I am fairly new to unittests, so are there any pointers that I could follow?
Thanks!

**Comment by javabster** [comment]
Great thanks @tgag17!

If you follow the steps I commented above that should give you a good outline of what to do, there are examples of snapshot tests  in `test_circuit_matplotlib_drawer.py` that you can use as reference. `test_circuit_matplotlib_drawer.py` is testing the functions in `qiskit/visualization/circuits_visualization.py`, but your file `test_counts_matplotlib_drawer.py` should test the functions in `qiskit/visualization/counts_visualization.py`.

As for general pointers make sure you name your tests appropriately, using the prefix `test_` and add a short description. Try to think of a range of test cases (e.g. trying out different arguments), not just the obvious ones.

**Comment by tgag17** [comment]
Hi! Sorry, I was a bit preoccupied with some work and so wasn't able to make a PR. I will make at least a draft by this weekend. Thank you!

**Comment by 1ucian0** [comment]
un-assigning due lack of activity.

**Comment by epelaaez** [comment]
Hi @1ucian0, if this still needs to be done, I'd be happy to do it

**Comment by 1ucian0** [comment]
assgined @epelaaez !

**Comment by epelaaez** [comment]
Just before starting with this, I noticed that in `qiskit-terra/test/ipynb/mpl/graph/test_graph_matplotlib_drawer.py` there are tests for `plot_histogram` which is in `counts_visualization.py`. The only other functions on `counts_visualization.py` are `_plot_histogram_data` and `hamming_distance`, but I guess these are indirectly tested when testing `plot_histogram`.

If I understand correctly, what I have to do is add more tests for `plot_histogram` with different arguments, as right now the only tests is:

```python
def test_plot_histogram(self):
        """for testing the plot_histogram"""
        # specifing counts because we do not want oscillation of
        # result until a changes is made to plot_histogram

        counts = {"11": 500, "00": 500}

        self.graph_count_drawer(counts, filename="histogram.png")
```

I just want to make sure this is the right approach. And one other thing, should this be done in another file like stated in @javabster comment a while ago or should I add these new tests into `test_graph_matplotlib_drawer.py`? Thanks!

**Comment by 1ucian0** [comment]
you are right! the coverage from the original issue is outdated, the current coverage percentage is 71%. There are some branches to cover yet, if you are up to it. Otherwise, I think the issue can be closed.

https://coveralls.io/builds/43213044/source?filename=qiskit%2Fvisualization%2Fcounts_visualization.py

**Comment by epelaaez** [comment]
Sure, I'll be happy to work on this. But it'd be great if I can get some pointers about the specific things that need coverage. I looked at the link you provided, but I'm not completely sure how to interpret it. Do the lines highlighted in red mean that those are not called on any tests?

Also, new tests would go into `test_graph_matplotlib_drawer.py`, right?

**Comment by 1ucian0** [comment]
Yes, the lines in red means that the tests do no longer execute them. Search for the condition branches and parameterize the call parameters for those branches.

Also yes. Add the test to the same file. Thank you!

**Comment by epelaaez** [comment]
I added a simple test so I could make sure everything worked the way I think it does. However, it doesn't go as expected. Even though I pushed the change into my fork of the repo, my new test and other tests are not running in the binder. To make sure this was not only happening on my side, I went ahead and ran the binder for the orignal repo. However, some tests are also not running there. 

In `test_graph_matplotlib_drawer.py`, there are 19 tests. However, only 14 are showing up when running the tests in binder. Also, some errors seem to happen, but I'm not able to get the whole traceback for them. Since this is happening when testing the `main` branch from the original repo, I don't think it has to do with something I've done.

**Comment by javabster** [comment]
Hi @epelaaez it's hard to guess what the issue is you're facing without seeing how you've written your test or how you've been trying to run them, could you share a code snippet of what you've done? Or even better open a draft PR and link it to this issue so we can see the code changes you're making. When I run binder for main in my fork I can see all 19 test results 😄

**Comment by epelaaez** [comment]
Hi @javabster, thanks for the comment! I've opened #7096 that contains the small sample test I did. 

When I go into https://mybinder.org/v2/gh/epelaaez/qiskit-terra/patch-1?urlpath=apps/test/ipynb/mpl_tester.ipynb, I only see 14 tests instead of the 20 in that file. The same happens when I go into https://mybinder.org/v2/gh/qiskit/qiskit-terra/main?urlpath=apps/test/ipynb/mpl_tester.ipynb. Maybe I'm using an old link? For reference, I tried in Chrome and Opera browsers.
